src/fges.jl

```diff
@@ -0,0 +1,536 @@
+==(a::T, b::T) where T <: ParseData = a.data == b.data
+
+
+####################################################################
+# Main Entry point for the Algorithm
+####################################################################
+
+"""
+    fges(data; penalty = 1.0, verbose=false)
+
+Compute a causal graph for the given observed data.
+"""
+function fges(data; penalty = 1.0, verbose=false)
+
+    # data type / precision
+    scoreT = eltype(data)
+
+    # Parse the inputted data
+    dataParsed = ParseData(data, penalty)
+
+    # Create an empty graph with one node for each feature
+    g = DiGraph(dataParsed.numFeatures)
+
+    return fges_internal(g, scoreT, dataParsed; penalty, verbose)
+end
+
+function fges_internal(g, scoreT, data; penalty = 1.0, verbose=false)
+
+    verbose && println("Start forward search")
+    # Perform the forward search 
+    
+    ges_search!(g, Step{Int, scoreT}(), data, Insert!, verbose)
+
+    verbose && println("Start backward search")
+    # Perform the backward search 
+    ges_search!(g, Step{Int, scoreT}(), data, Delete!, verbose)
+    
+    return g # Return the graph
+
+end
+
+####################################################################
+# Insert and Delete Operators
+####################################################################
+
+function Insert!(g, newStep::Step) 
+    edge = newStep.edge
+    T = newStep.subset
+
+    # Add a directed edge x→y
+    add_edge!(g,edge)
+    
+    # Orient all edges incident into child node
+    y = dst(edge)
+    for t ∈ T
+        orientedge!(g,t,y) #t→y
+    end
+
+    return nothing
+end
+
+
+function Delete!(g, newStep::Step)
+    edge = newStep.edge
+    H = newStep.subset
+
+    # Remove directed and unidirected edges (x→y and x-y)
+    rem_edge!(g, edge)
+    rem_edge!(g, reverse(edge))
+    
+    # Orient all vertices in H toward x and y
+    (x, y) = src(edge), dst(edge)
+    for h ∈ H
+        orientedge!(g,x,h) #x→h
+        orientedge!(g,y,h) #y→h
+    end
+
+    return nothing
+end
+
+####################################################################
+# Forward and Backward search
+####################################################################
+
+function ges_search!(g, newStep, data, operator, verbose)
+
+    # Continually add/remove edges to the graph until the score stops increasing
+    while true
+        
+        # Get the new best step (depends on if we're inserting or deleting)
+        if operator == Insert!
+            findNextEquivClass!(newStep, data, g, findBestInsert, verbose)
+        else
+            findNextEquivClass!(newStep, data, g, findBestDelete, verbose)
+        end
+        
+        
+        # If the score did not improve...
+        if newStep.Δscore ≤ 0
+            #...stop searching
+            break
+        end
+        
+        # Use the insert or delete operator update the graph
+        operator(g, newStep)
+        
+        # Covert the PDAG to a complete PDAG
```

Comment:
Maybe we should add a general function for this as its also needed in PC algorithm. I could open a PR. We wrote some code in the context of this paper (https://arxiv.org/abs/2302.14386), which is available here: https://github.com/Malte311/PdagOrientations

Reply:
This is already contained in some way in my implementation of the PC algorithm and I have made the code more modular. I'll point it out in a comment.

Reply:
Meek https://arxiv.org/pdf/1302.4972.pdf writes: "Chickering (1995) also gives an algorithm to find the maxi­ mally oriented graph from a pattern; this algorithm is more complicated but more efficient than a naive implementa tion of the method described above." Not very clear which of the two algorithms (Meeks orientation rules or Chickering 1995) is "this algorithm" and which is "the method described above."

Reply:
Chickerings is "this algorithm" (more complicated but more efficient) and Meeks "the method described above".

Meek's approach is:
- Take the PDAG and apply R1-R4 repeatedly until no rule applies. [To obtain the CPDAG from a pattern R1-R3 suffice]

Chickerings approach is:
- Take the PDAG and find a DAG with same skeleton and v-structures. (This can be done e.g. with this algorithm by Dor and Tarsi: https://ftp.cs.ucla.edu/pub/stat_ser/r185-dor-tarsi.pdf)
- From the DAG construct the CPDAG (which is possible in linear-time, we also have this implemented ```cpdag(g)```).

For us Chickerings approach would be an option because the complicated part (from an implementation perspective) is DAG-to-CPDAG which we already have. A straightforward implementation of PDAG to DAG by Dor and Tarsi are just a few lines.

Two comments:
- for speed of PC and GES this will likely not matter too much overall, other parts are more costly
- in case of PC the pattern learned from data might not have a DAG with same skeleton and v-structures. Then, it is not clear how to use Chickerings approach. (It is also not clear what the Meek rules really do in this case, I have some thoughts on this, but this PR is not the place...)

Reply:
This should be a dashed line for Rule 4 in your paper? 
![IMG_0543](https://github.com/mschauer/CausalInference.jl/assets/1923437/e1ef1e60-8fe2-4475-b586-a2ebb090d6a4)

Reply:
Dashed as in "it does not matter which way it is oriented"?

I think it's actually not necessary:
- if we have a -> c, then a -> b follows by R2 (because of a -> c -> b)
- if we have a <- c, then d -> a follows by R2 (because of d -> c -> a) and consequently a -> b by R1 (because of d -> a - b)

So it should suffices to test only for an undirected edge between a and c.

Reply:
This should go into some kind of  developer docs